Improved Mod Manager UI

HUSSAR-mtrela opened the pull request:
Improved UI and added new features to the Mod Manager:

- mod sorting
- mod searching
- save/load a list of favorite mods
- compact/expanded view of mod list
- links to mod source/websites
- added optional styling arguments to tooltip functions

Mod Manager in compact view:
![image](https://user-images.githubusercontent.com/11665978/177224394-f151249f-fbf8-4474-a62a-555967115c07.png)

Mod Manager in expanded view: 
![image](https://user-images.githubusercontent.com/11665978/177224083-88aa4b15-1a59-4306-a33c-c68ad132f074.png)


Searching for mods with a "black" keyword:
![image](https://user-images.githubusercontent.com/11665978/177222589-fcc0fc0d-a801-4e22-b394-7d0c8fa25ff8.png)

Sorting mods by their name:
![image](https://user-images.githubusercontent.com/11665978/177222977-d1ad1ff3-6ee0-4b79-bd68-f5b4e6537480.png)

Save Favorite Mods:
![image](https://user-images.githubusercontent.com/11665978/177223946-95d4b64b-557c-4926-b949-da48833aa286.png)

Load Favorite Mods:
![image](https://user-images.githubusercontent.com/11665978/177223811-dd8a7d46-27c5-4f0b-a3c8-452c119dd058.png)

Mod Manager show:
-  links to mod info website (if mod_info.lua has url setting)
-  links to github website (if mod_info.lua has github setting)

![image](https://user-images.githubusercontent.com/11665978/177224159-ce04338b-cd41-4ab8-abea-a1c35118ffd5.png)

Garanas commented on lua/ui/game/tooltip.lua:
```diff
@@ -20,39 +20,42 @@ local Keymapping = import('/lua/keymap/defaultKeyMap.lua').defaultKeyMap
+-- creates a tooltip box from ID table and with optional parameters
+-- @param - ID is a table, e.g. { text = 'tooltip header', body = 'tooltip description' } 
+-- @param - extendedBool indicates whether to just create tooltip header or also tooltip description
+-- @param - width is optional width of tooltip or it is auto calculated based on length of header/description
```
The format of the new annotations is `--- @param name type description`, as an example:
`---@param width number Optional width of tooltip or it is auto calculated based on length of header/description`.

See also: https://github.com/FAForever/fa/pull/4009/files as an example

HUSSAR-mtrela replied:
ok, I will correct API comments

Garanas commented on lua/ui/game/tooltip.lua:
```diff
@@ -193,34 +218,43 @@ function CreateExtendedToolTip(parent, text, desc, width)
             if text == nil then
-                tooltip.desc[1].Top:Set(tooltip.Top)
-                tooltip.desc[1].Left:Set(tooltip.Left)
+--                tooltip.desc[1].Top:Set(tooltip.Top)
```
Prefer to remove code instead of commenting it

Garanas reviewed (commented):
The difference in your screenshots (where the star icon becomes green) doesn't happen:

![image](https://user-images.githubusercontent.com/15778155/177264756-4004433f-bf2e-481f-b68c-49044d34a645.png)

There's some empty space over at the end, we can take that out

![image](https://user-images.githubusercontent.com/15778155/177264818-7ac20751-7ea2-4535-890b-2abfb0621029.png)

Great work in general - very exciting!

Garanas commented:
Done 👍

HUSSAR-mtrela commented:
@Garanas I fixed all issues that you found:
- validating domain of URLs before calling OpenURL()
- styling of favorite icon when Random faction is selected
- formatting of API comments in tooltip.lua
- removed dead/commented out code

Garanas reviewed (changes requested):
![image](https://user-images.githubusercontent.com/15778155/178423292-d12dae08-7d2d-4466-855e-48094e4126a8.png)

The list count doesn't adjust when you toggle filters on / off

![image](https://user-images.githubusercontent.com/15778155/178423435-97a5e0fe-c240-4737-9a4c-4a45847638fe.png)

When you toggle on the favorite mods the list resets, and you end up at the top again. That feels odd

The default color of the stars should be darker, to have a significant difference when you've chosen a random faction. I tried `ff434343`, that works good

@HUSSAR-mtrela Do you think you can find the time to do this in the next few days? I'd like to include this with the release, but for that I'd need to merge it sooner than later 😄

HUSSAR-mtrela commented:
@Garanas alright, I fixed scroll bar when mods are filtered and I used your color for unchecked favorite toggle. Also, I fixed counter for blacklisted mods and mods missing dependency. 

However, I did not change resting of mod list to top after toggling on the favorite mods because the same happens when filtering mods since both filtering mods and loading favorite mods are using the same function for refreshing mods list. Changing this behavior would be risky and it would add additional complexity. Ideally, this change would be much easier done when I replace (in another PR/release) the custom scroll control used in the Mod Manager with existing Grid control which has built-in scrolling.

Garanas commented:
Sounds good - and it is looking good. Great work!

4z0t commented:
Really like new look, however:
1. I'd really appreciate if it is done in new style, so, there is a class that handles it, lines  has class too and so on
2. Use Layouter instead of LayoutHelpers
3. Star's background doesnt match with ingame one 
![image](https://user-images.githubusercontent.com/84857900/179385587-eb2e3d1a-5e82-4c92-989b-367ce27a01af.png)

4z0t commented on lua/ui/lobby/ModsManager.lua:
```diff
@@ -127,18 +162,19 @@ local modsTags = {
 function CreateDialog(parent, isHost, availableMods, saveBehaviour)
```
must be a Class

4z0t commented on lua/ui/lobby/ModsManager.lua:
```diff
@@ -155,31 +191,109 @@ function CreateDialog(parent, isHost, availableMods, saveBehaviour)
+
+    -- creating a toggle for activating favorite mods
+    local favoriteModsToggle = CreateFavoriteButton(dialogContent, 25, false, true)
+    LayoutHelpers.AtLeftIn(favoriteModsToggle.Icon, dialogContent, 18)
```
forget LayoutHelpers, use Layouter

Garanas replied:
This PR builds on legacy code, it is not a reimplementation. For now, this is fine
